metis-media-service/src/main/java/eu/europeana/metis/mediaprocessing/RdfDeserializerImpl.java

```diff
@@ -6,50 +6,134 @@
 import eu.europeana.metis.mediaprocessing.model.EnrichedRdf;
 import eu.europeana.metis.mediaprocessing.model.EnrichedRdfImpl;
 import eu.europeana.metis.mediaprocessing.model.RdfResourceEntry;
-import eu.europeana.metis.mediaprocessing.model.RdfWrapper;
 import eu.europeana.metis.mediaprocessing.model.UrlType;
 import java.io.ByteArrayInputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Optional;
 import java.util.Set;
 import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import javax.xml.XMLConstants;
+import javax.xml.namespace.NamespaceContext;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.xpath.XPath;
+import javax.xml.xpath.XPathConstants;
+import javax.xml.xpath.XPathExpression;
+import javax.xml.xpath.XPathExpressionException;
+import javax.xml.xpath.XPathFactory;
 import org.jibx.runtime.IUnmarshallingContext;
 import org.jibx.runtime.JiBXException;
+import org.w3c.dom.Document;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+import org.xml.sax.SAXException;
 
 /**
- * This object implements RDF deserialization functionality.
+ * This implements RDF deserialization functionality. The code that obtains the individual resources
+ * does not assume that we can convert the record to an EDM internal format. Link checking must also
+ * run on EDM external. We therefore use XPath expressions to obtain the required data.
  */
 class RdfDeserializerImpl implements RdfDeserializer {
 
   private final IUnmarshallingContext context;
 
+  private final XPathExpression getObjectExpression;
+  private final XPathExpression getHasViewExpression;
+  private final XPathExpression getIsShownAtExpression;
+  private final XPathExpression getIsShownByExpression;
+
   /**
    * Constructor.
    *
    * @throws RdfConverterException In case something went wrong constructing this object.
    */
   RdfDeserializerImpl() throws RdfConverterException {
+
+    // Create the unmarshalling context
     try {
       context = RdfBindingFactoryProvider.getBindingFactory().createUnmarshallingContext();
     } catch (JiBXException e) {
       throw new RdfConverterException("Problem creating deserializer.", e);
     }
+
+    // Create xpath engine and set namespace context.
+    final XPath xPath = XPathFactory.newInstance().newXPath();
```

Comment:
I think we should be using the more robust Saxon library rather than the default..

Reply:
Perhaps. But that is not necessarily a straight-forward change (by the looks of it). This is code I just copied over, so it's not like I chose this way of processing here. It would be a new change and one that also need to happen in other parts of the code. Perhaps we can create a maintenance ticket for this to separate this from the work here?

Reply:
Same here as above:

"Ok let's create a separate ticket to possibly improve that.
But if you are saying that it's design to be instantiated per worker then we should definitely use Saxon because the worker threads can be on the same JVM and the default implementation has some static fields which can cause an issue..(Saxon implementation does not contain those static fields)
And/Or synchronize those..
We found an issue like that recently(I forgot the component, I think transformation..)"

Reply:
Same as above:
I made some changes to make things load lazily and in a thread-safe way. Could you create the ticket for what's left to do?